### DISCUSSION TITLE
How to configure Time_Format to parse UNIX timestamp in miliseconds?

### DISCUSSION THREAD

**Discussion by gajus** [fluent/fluent-bit#6502] [Q&A]
I am using [Roarr](https://github.com/gajus/roarr) JSON logger, which produces logs in the following format:

```json
{
  "context": {
    "application": "contra-web-app",
    "namespace": "server",
    "logLevel": 10
  },
  "message": "render page time 23 ms",
  "sequence": "1285",
  "time": 1670267708402,
  "version": "2.0.0",
  "severity": "DEBUG"
}
```

It is not clear when reading [parser configuration](https://docs.fluentbit.io/manual/pipeline/parsers/configuring-parser) how to configure `Time_Format` given that it asks to specify format compatible with `strptime`, and the latter does not have format for milliseconds.

I see that the documentation includes section about fractional seconds, but it only mentions nanoseconds and not milliseconds:

> In addition, we extended our time resolution to support fractional seconds like 2017-05-17T15:44:31**.187512963**Z. Since Fluent Bit v0.12 we have full support for nanoseconds resolution, the %L format option for Time_Format is provided as a way to indicate that content must be interpreted as fractional seconds.

I tried `%s%L`, but didn't have the desired effect.

https://docs.fluentbit.io/manual/pipeline/parsers/configuring-parser

**Comment by fovgrubby128** [comment]
bump up! Not sure how you solve it in the end...

**Comment by jmtt89** [comment]
it's not posible, the easiest fix is modifying you server to generate date with toISOString() and use `%Y-%m-%dT%H:%M:%S.%LZ`

another option i found is use a lua script for preformat the time field.

**Comment by selmison** [comment]
A way is: `%s.%L` 

Notice this [comment](https://github.com/fluent/fluent-bit/blob/393fe542e4602fc87981bca99b4d19c8ac5fbb8c/src/flb_parser.c#L289):

> Since strptime(3) does not support fractional seconds, this
> requires a workaround/hack in our parser. This is a known
> issue and addressed in different ways in other languages.
> 
> The following links are a good reference:
> - http://stackoverflow.com/questions/7114690/how-to-parse-syslog-timestamp
> - http://code.activestate.com/lists/python-list/521885

**Reply by lake2** [reply]
> A way is: `%s.%L`
> 
> Notice this [comment](https://github.com/fluent/fluent-bit/blob/393fe542e4602fc87981bca99b4d19c8ac5fbb8c/src/flb_parser.c#L289):
> 
> > Since strptime(3) does not support fractional seconds, this
> > requires a workaround/hack in our parser. This is a known
> > issue and addressed in different ways in other languages.
> > The following links are a good reference:
> > 
> > * http://stackoverflow.com/questions/7114690/how-to-parse-syslog-timestamp
> > * http://code.activestate.com/lists/python-list/521885

It's not working

**Comment by lake2** [comment]
same issue +1
